File: frontend/src/types/fabric.ts
/**
 * Fabric Recommendation Types - Story 2.3
 *
 * TypeScript types and Zod schemas for fabric recommendation API.
 * Uses snake_case for API field names to match Backend SSOT.
 */

import { z } from "zod";

// ===== Fabric Property Schema =====

export const fabricPropertySchema = z.object({
  do_ru: z.string(),
  do_day: z.string(),
  do_co_dan: z.string(),
  do_bong: z.string(),
  kha_nang_giu_phom: z.string(),
});

export type FabricProperty = z.infer<typeof fabricPropertySchema>;

// ===== Fabric Response Schema =====

export const fabricResponseSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string(),
  image_url: z.string().nullable(),
  properties: fabricPropertySchema,
  compatibility_score: z.number().min(0).max(100),
  compatibility_label: z.string(),
});

export type FabricResponse = z.infer<typeof fabricResponseSchema>;

// ===== Fabric Recommendation Response Schema =====

export const fabricRecommendationResponseSchema = z.object({
  pillar_id: z.string(),
  fabrics: z.array(fabricResponseSchema),
  total: z.number(),
});

export type FabricRecommendationResponse = z.infer<typeof fabricRecommendationResponseSchema>;

// ===== Vietnamese property labels (for UI display) =====

export const FABRIC_PROPERTY_LABELS: Record<keyof FabricProperty, string> = {
  do_ru: "Độ rủ",
  do_day: "Độ dày",
  do_co_dan: "Độ co dãn",
  do_bong: "Độ bóng",
  kha_nang_giu_phom: "Giữ phom",
};

export const FABRIC_LEVEL_LABELS: Record<string, string> = {
  none: "Không",
  low: "Thấp",
  medium: "Trung bình",
  high: "Cao",
  light: "Mỏng",
  heavy: "Dày",
};

File: backend/src/models/fabric.py
"""Pydantic schemas for Fabric Recommendations - Story 2.3.

Defines models for fabric properties, individual fabric details,
and recommendation response with compatibility scores.
Uses 100% Vietnamese tailoring terminology per NFR11.
"""

from pydantic import BaseModel, Field


class FabricProperty(BaseModel):
    """Physical and aesthetic properties of a fabric type.

    Properties are expressed as Vietnamese textile terminology
    with values: "none", "low", "medium", "high".
    """

    do_ru: str = Field(..., description="Độ rủ — drape quality (none/low/medium/high)")
    do_day: str = Field(..., description="Độ dày — thickness (light/medium/heavy)")
    do_co_dan: str = Field(..., description="Độ co dãn — stretch (none/low/medium/high)")
    do_bong: str = Field(..., description="Độ bóng — shine/sheen (none/low/medium/high)")
    kha_nang_giu_phom: str = Field(
        ..., description="Khả năng giữ phom — shape retention (low/medium/high)"
    )

    model_config = {"from_attributes": True}


class FabricResponse(BaseModel):
    """Response schema for a single fabric recommendation (Story 2.3).

    Represents a fabric type with its properties and compatibility score.
    """

    id: str = Field(..., description="Unique fabric identifier")
    name: str = Field(..., description="Tên vải bằng tiếng Việt")
    description: str = Field(..., description="Mô tả đặc tính vải")
    image_url: str | None = Field(None, description="URL hình ảnh texture")
    properties: FabricProperty = Field(..., description="Đặc tính vật lý của vải")
    compatibility_score: float = Field(
        ..., ge=0.0, le=100.0, description="Mức độ phù hợp (0-100)"
    )
    compatibility_label: str = Field(
        ..., description="Nhãn phù hợp: Rất phù hợp / Phù hợp / Có thể dùng"
    )

    model_config = {"from_attributes": True}


class FabricRecommendationResponse(BaseModel):
    """Response schema for fabric recommendation list (Story 2.3)."""

    pillar_id: str = Field(..., description="ID phong cách đã chọn")
    fabrics: list[FabricResponse] = Field(
        default_factory=list, description="Danh sách vải gợi ý, sắp xếp theo compatibility"
    )
    total: int = Field(0, description="Tổng số loại vải gợi ý")

    model_config = {"from_attributes": True}
